tests/ci/cdk/pipeline/codebuild_batch_step.py

```diff
@@ -0,0 +1,100 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0 OR ISC
+import builtins
+import re
+import typing
+from typing import Mapping
+
+import jsii
+from aws_cdk import (
+    Duration,
+    pipelines,
+    aws_codepipeline_actions as cp_actions,
+    aws_codebuild as codebuild,
+    aws_codepipeline as codepipeline,
+    aws_iam as iam,
+)
+
+
+@jsii.implements(pipelines.ICodePipelineActionFactory)
+class CodeBuildBatchStep(pipelines.Step):
+    """
+    Create a CodeBuildBatchStep given shell commands and batch build settings.
+
+    :param id: The id of the step.
+    :param input: The input file set producer.
+    :param action_name: Name of the action produced by this step.
+    :param commands: The CodeBuild commands to be run.
+    :param partial_batch_build_spec: The batch build settings for the project.
+    :param role: The role to use for the CodeBuild project.
+    :param timeout: Timeout of the batch build project, in minutes.
+    :param env: The environment variables to use for the CodeBuild project.
+
+    :return: A new CodeBuildBatchStep.
+    """
+
+    def __init__(
+        self,
+        id,
+        input: pipelines.FileSet,
+        action_name: str,
+        commands: typing.List[str],
+        partial_batch_build_spec: typing.Mapping[builtins.str, typing.Any],
+        role: iam.Role,
+        timeout: int = 300,
+        project_description: str = None,
+        env: typing.Optional[typing.Mapping[str, str]] = None,
+    ):
+        super().__init__(id)
+
+        self._discover_referenced_outputs(env)
```

Comment:
This seems to be another private-y function you are invoking from `pipelines.Step`, is this really what the CDK expects a user to do?

Reply:
Yes. For a custom step to use other step's outputs, it needs to call `_discover_referenced_outputs`. The documentation mentions this
https://github.com/aws/aws-cdk/blob/be6210f246b97befcdc9446862e991071738008d/packages/aws-cdk-lib/pipelines/README.md?plain=1#L852-L860